### PR TITLE
First step to test  how block producers are selected without chunk-on…

### PR DESCRIPTION
**Pull request by yordanmadzhunkov**
…ly producers

https://github.com/near/nearcore/issues/4813

### CODE REVIEW

File: chain/epoch_manager/src/validator_selection.rs [hunk 1]
```diff
@@ -487,6 +488,145 @@ mod tests {
...
+            );
+            prev_stake = v.stake();
+        }
+
+        assert_eq!(epoch_info.block_producers_settlement().len() as u64, num_block_producer_seats);
+        assert_eq!(epoch_info.chunk_producers_settlement().len() as u64, num_shards);
+    }
+
+    #[test]
+    fn test_block_producer_sampling_with_100_block_producers() {
```
Comment (bowenwang1996) [inline chain/epoch_manager/src/validator_selection.rs:566]: Probably makes sense to write something that can take some sort of config as input (similar to how runtime-tester is implemented) to make writing tests easier

Comment (yordanmadzhunkov) [reply chain/epoch_manager/src/validator_selection.rs:583]: Done :)

**Comment by yordanmadzhunkov** [comment]
Useful for testing:

`cargo test --workspace --features nightly_protocol,nightly_protocol_features --jobs 6 test_block_producers_with_sampling -- 2>&1 | grep 'Producing block'
`

feel free to remove the grep, if you want to read the entire log

File: integration-tests/tests/client/process_blocks.rs [hunk 2]
```diff
@@ -3891,4 +3891,76 @@ mod contract_precompilation_tests {
...
+            env.clients[producer_id].process_tx(tx, false, false);
+            // Check that we are were called at the block that we are producer for.
+            let block =
+                env.clients[producer_id].produce_block(next_block_height as u64).unwrap().unwrap();
+            for i in 0..num_clients {
+                let provenance =
+                    if i == producer_id { Provenance::PRODUCED } else { Provenance::SYNC };
+                let _ = env.clients[producer_id].process_block(block.clone(), provenance);
+            }
+            prev_block_hash = block.hash().clone();
```
[the remark this answers is not in the document]
Comment (yordanmadzhunkov) [reply integration-tests/tests/client/process_blocks.rs]: I added such logic with one transaction per block. Unfortunetly, some of this transactions got rejected and I can't understand why ...
```
Sep 24 18:07:52.250 DEBUG chunks: Reconstructed and decoded chunk HDp4CEx4UDVnwTCTdDZDqmVHMi3TXjoVbxefvhnBgBX5, encoded length was 185, num txs: 1, I'm Some(AccountId("test3"))    
Sep 24 18:07:52.262 DEBUG chunks: Reconstructed and decoded chunk 6fxrPtLKNfj2EtQoKHMZnnyW8UhKmztaxf81ZeTaZmTq, encoded length was 8, num txs: 0, I'm Some(AccountId("test0"))    
Sep 24 18:07:52.270 DEBUG chunks: Reconstructed and decoded chunk F89CavLi9UqXppeHdaMFxoh255Cu23H7C3zQzDcEfeym, encoded length was 8, num txs: 0, I'm Some(AccountId("test2"))    
Sep 24 18:07:52.280 DEBUG chunks: Reconstructed and decoded chunk 29zT97qFa7PUCkvc7phXUUrzHaUJARsqJCGe1QkyZaSa, encoded length was 8, num txs: 0, I'm Some(AccountId("test1"))    
Sep 24 18:07:52.289 DEBUG chunks: Reconstructed and decoded chunk Y1rDB8kWGyUDuRkz63w5tPcMAQD83NyH3gtjVaCZPcS, encoded length was 8, num txs: 0, I'm Some(AccountId("test0"))    
Sep 24 18:07:52.299 DEBUG chunks: Reconstructed and decoded chunk SBpMXqRhGTdRGceGud2YazHuoaieWpm6yDK7Adu4atG, encoded length was 8, num txs: 0, I'm Some(AccountId("test1"))    
Sep 24 18:07:52.309 DEBUG chunks: Reconstructed and decoded chunk G9s1vXKaoThtrVqJWbYT3YR9bWkM8Hs7WxqeCqW9m92z, encoded length was 185, num txs: 1, I'm Some(AccountId("test2"))    
Sep 24 18:07:52.318 DEBUG chunks: Reconstructed and decoded chunk GpnVKxW4n6Vp1CYwas5ydgmxqKciPT5JjgZjrFgAUZN, encoded length was 8, num txs: 0, I'm Some(AccountId("test0"))    
Sep 24 18:07:52.326 DEBUG chunks: Reconstructed and decoded chunk 4PaBnxiM3fxrBx5qwmz34pb6Q7gCKBMkAwj4Pj4G9Xxd, encoded length was 185, num txs: 1, I'm Some(AccountId("test2"))    
Sep 24 18:07:52.335 DEBUG chunks: Reconstructed and decoded chunk 8PmiM5zQ4uqkiyFwPB1WAJMvAEa7UDkGcXPPcg59Kjvv, encoded length was 185, num txs: 1, I'm Some(AccountId("test2"))    
```

File: integration-tests/tests/client/process_blocks.rs [hunk 3]
```diff
@@ -3891,4 +3891,111 @@ mod contract_precompilation_tests {
...
+                //distribute chunks
+                client
+                    .shards_mgr
+                    .distribute_encoded_chunk(
+                        encoded_chunk,
+                        merkle_paths,
+                        receipts,
+                        &mut chain_store,
+                    )
+                    .unwrap();
```
Comment (bowenwang1996) [inline integration-tests/tests/client/process_blocks.rs:3987]: This is certainly moving in the right direction. However, I am not sure that it is correct. Here we distribute the chunk parts but I don't see any code that enables other block producers to receive them. You might have to change network adapter to make sure that chunk parts are delivered.

File: integration-tests/tests/client/process_blocks.rs [hunk 4]
```diff
@@ -3959,41 +3965,98 @@ mod contract_precompilation_tests {
                 .runtime_adapter
                 .account_id_to_shard_id(&signer.account_id, &epoch_id)
                 .unwrap();
-            // every client should know who is next chunk producer
+            // every     client should know who is next chunk producer
```
Comment (bowenwang1996) [inline integration-tests/tests/client/process_blocks.rs:3968]: what happened here

File: chain/chain/src/types.rs [hunk 5]
```diff
@@ -737,6 +737,19 @@ pub trait RuntimeAdapter: Send + Sync {
             })
             .collect()
     }
+
+    fn build_receipts_hashes2(
```
Comment (bowenwang1996) [inline chain/chain/src/types.rs:741]: this name is not very descriptive :(

File: integration-tests/tests/client/process_blocks.rs (hunk 4 above)
Comment (yordanmadzhunkov) [reply integration-tests/tests/client/process_blocks.rs:3968]: Looks like I press tab by accident :)

File: chain/chain/src/types.rs (hunk 5 above)
Comment (yordanmadzhunkov) [reply chain/chain/src/types.rs]: I agree. I will refactor the code

File: chain/chain/src/types.rs [hunk 6]
```diff
@@ -748,11 +748,12 @@ pub trait RuntimeAdapter: Send + Sync {
...
-        &self,
-        receipts: &[Receipt],
-        shard_layout: &ShardLayout,
-    ) -> Vec<CryptoHash> {
+    fn build_receipts_hashes(&self, receipts: &[Receipt], epoch_id: &EpochId) -> Vec<CryptoHash> {
+        let shard_layout = //{
+            //let epoch_id = self.get_epoch_id_from_prev_block(&prev_block_hash).unwrap();
+            self.get_shard_layout(&epoch_id).unwrap();
+        //};
+        //self.build_receipts_hashes(receipts, &shard_layout)
```
Comment (matklad) [inline chain/chain/src/types.rs:756]: Some left-over comments? I think it's best to avoid commented-out code in the repository, even if it's only temporarily. Such comments tend to stay for a long time more often than not.

File: chain/chunks/src/lib.rs [hunk 7]
```diff
@@ -1623,6 +1624,21 @@ impl ShardsManager {
         Ok(())
     }
 
+    pub fn build_block_producer_mapping(
```
Comment (matklad) [inline chain/chunks/src/lib.rs:1627]: This is test-only, right? Can we implement this function in a test, instead of it being publicly visible everywhere?

File: chain/chunks/src/lib.rs [hunk 8]
```diff
@@ -1623,6 +1624,21 @@ impl ShardsManager {
...
+        &self,
+        prev_block_hash: &CryptoHash,
+    ) -> HashMap<AccountId, Vec<u64>> {
+        let mut block_producer_mapping = HashMap::new();
+
+        for part_ord in 0..self.runtime_adapter.num_total_parts() {
+            let part_ord = part_ord as u64;
+            let to_whom = self.runtime_adapter.get_part_owner(&prev_block_hash, part_ord).unwrap();
+            let entry = block_producer_mapping.entry(to_whom).or_insert_with(Vec::new);
+            entry.push(part_ord);
```
Comment (matklad) [inline chain/chunks/src/lib.rs:1637]:
```suggestion
            block_producer_mapping.entry(to_whom)
                .or_default()
                .push(part_ord);
```

File: chain/chunks/src/lib.rs [hunk 9]
```diff
@@ -1623,6 +1624,21 @@ impl ShardsManager {
         Ok(())
     }
 
+    pub fn build_block_producer_mapping(
+        &self,
+        prev_block_hash: &CryptoHash,
+    ) -> HashMap<AccountId, Vec<u64>> {
+        let mut block_producer_mapping = HashMap::new();
+
+        for part_ord in 0..self.runtime_adapter.num_total_parts() {
```
[the remark this answers is not in the document]
Comment (matklad) [reply chain/chunks/src/lib.rs:1633]: Feels like this loop can be simplified to a `.collect` call. No strong opinion here, but better to use `.collect` if that's possible and shorter.

File: integration-tests/tests/client/process_blocks.rs [hunk 10]
```diff
@@ -3938,4 +3942,181 @@ mod contract_precompilation_tests {
         // Check that contract is not cached for client 1 because of late state sync.
         assert!(caches[1].get(&contract_key.0).unwrap().is_none());
     }
+
+    #[test]
+    fn test_block_producers_with_sampling() {
```
Comment (matklad) [inline integration-tests/tests/client/process_blocks.rs:3947]: Is it intended that this function is inside `contract_precompilation_tests`? Feels unrelated to contract precompilation.

File: chain/chunks/src/lib.rs (hunk 7 above)
Comment (yordanmadzhunkov) [reply chain/chunks/src/lib.rs:1627]: Yes, right now it's only used in test code. For me, this piece of code belongs to ShardsManager because it uses only ShardManager data members to compute the result. If you think there is a low probability this method to be reused, I will move it to test code.

Comment (matklad) [reply chain/chunks/src/lib.rs:1627]: No super strong opinion here, but I'd define

```rust
fn build_block_producer_mapping(
  shard_manager: &ShardManager,
  prev_block_hash: &CryptoHash,
)
```

in the test code. 

"where the thing is used from" is somewhat more important than "what is needed to implement the thing". public inherent methods should be used for generally re-usable functionality. If you need something in one specific case, better to make it a plain function.

**Comment by bowenwang1996** [comment]
@yordanmadzhunkov if the PR is ready for review please re-request reviews. Thanks!

**Comment by mzhangmzz** [comment]
what's the state for this PR? Is there plan to continue working on it or abandoning it?

**Comment by bowenwang1996** [comment]
@yordanmadzhunkov ?

**Comment by akhi3030** [comment]
This PR looks abandoned.
